packages/toolboxes/substrate/src/toolbox/baseSubstrateToolbox.ts

```diff
@@ -147,28 +166,41 @@ export const BaseSubstrateToolbox = ({
   api: ApiPromise;
   network: SubstrateNetwork;
   gasAsset: AssetValue;
-  signer: IKeyringPair;
+  signer: IKeyringPair | Signer;
 }) => ({
   api,
   network,
   decodeAddress,
   encodeAddress,
   createKeyring: (phrase: string) => createKeyring(phrase, network.prefix),
-  getAddress: (keyring: IKeyringPair = signer) => keyring.address,
+  getAddress: (keyring: IKeyringPair | Signer = signer) =>
+    isKeyringPair(keyring) ? keyring.address : undefined,
   createTransfer: ({ recipient, assetValue }: { recipient: string; assetValue: AssetValue }) =>
     createTransfer(api, { recipient, amount: assetValue.getBaseValue("number") }),
   getBalance: (address: string) => getBalance(api, gasAsset, address),
   validateAddress: (address: string) => validateAddress(address, network.prefix),
   transfer: (params: SubstrateTransferParams) => transfer(api, signer, params),
   estimateGasFee: (params: SubstrateTransferParams) =>
     estimateGasFee(api, signer, gasAsset, params),
-  sign: (tx: SubmittableExtrinsic<"promise">) => sign(signer, tx),
+  sign: (tx: SubmittableExtrinsic<"promise">) => {
+    if (isKeyringPair(signer)) {
+      return sign(signer, tx);
+    }
+    throw new Error("Signer does not have keyring pair capabilities required for signing.");
+  },
   broadcast: (tx: SubmittableExtrinsic<"promise">, callback?: Callback<ISubmittableResult>) =>
     broadcast(tx, callback),
   signAndBroadcast: (
     tx: SubmittableExtrinsic<"promise">,
     callback?: Callback<ISubmittableResult>,
-  ) => signAndBroadcast(signer, tx, callback),
+  ) => {
+    if (isKeyringPair(signer)) {
+      return signAndBroadcast(signer, tx, callback);
+    }
+    throw new Error(
+      "Signer does not have keyring pair capabilities required for signing and broadcasting.",
+    );
```

Comment:
Dont' use custom errors - this should be `SwapKitError` with proper error code

Reply:
fixed